### ISSUE TITLE
Project artifacts API

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#4881]
## Overview

There should be a way for a software project to declare its artifacts and how to build them with Dagger, in such a way that the artifacts can be discovered and produced on-demand via the Dagger API.

This is related to #4414, but focuses on a one narrow possible application of a project entrypoint: specifically *artifacts*.

## Design

I'm not sure how this would work, but here is a rough sketch to stimulate conversation.

###  Frontend API

API to discover and consume artifacts from the engine.

```graphql
extend type Project {
  artifacts: [...Artifact!]!
  artifact(name: String!): Artifact!
}

type Artifact {
  // Name of the artifact
  name: String!
  // Name of the extension that produced this extension
  provider: String!
  // Human-friendly description of the artifact
  description: String
  // Available versions of this artifact
  versions: [...ArtifactVersion!]!
  // Additional metadata
  labels: [...ArtifactLabel!]
}

type ArtifactVersion {
  // Version name
  version: String!
  // Contents of the artifact at that version
  contents: Directory!
  // Version-specific metadata
  labels: [...ArtifactLabel!]
}

type ArtifactLabel {
  key: String!
  value: String
}
```

### Backend API

API to expose artifacts to the engine. I'm honestly not sure how this would work. But I think it would involve the following:

1) A basic implementation of [extensions](#4796) that allows the controller code ("client") to be loaded and instrumented by the engine inside a container, rather than executed directly by the end user on the host.
2) SDK support for the controller code to register artifacts, with the most pleasant possible DX. Bonus points if the same controller code can somehow be executed directly *and* loaded as an extension, but I wouldn't say it's a must-have.
3) A new configuration file which allows loading certain extensions in the context of certain directories

### CLI

The CLI can be used to explore and consume artifacts directly from the command-line. This makes it easy for everyone in the team to experience the benefits of Dagger, without being familiar with it or having to know how to program it.

```console
$ dagger artifact list
...
```

```console
$ dagger artifact save
```

```console
$ dagger artifact publish
```

This is half-baked at best. Take this as an invitation to react and discuss.

### Docker registry bridge

This may be a crazy idea. But once there is a way for any projet to expose artifacts, and some of those artifacts are OCI-compatible containers... What if the dagger engine could act as an "virtual registry", where all OCI artifacts could be docker-pulled directly?

```console
$ dagger registry
Registry listening on localhost:4444
```

```
docker pull localhost:4444/backend:v1.3
```

The idea is that all artifacts are built on-demand. This could make local development much easier, compared to the recent guide we discussed @helderco @marcosnils @vikram-dagger :)

So that's it, thanks for reading! I would love to hear your thoughts.

cc @sipsma @vito @aluzzardi

**Comment by dolanor** [comment]
This is a very quick reaction to what I read, I haven't taken time to correctly to digest the idea.
We should have some kind of way to track platform as well, to know where we can use those artifacts.

```diff
type Artifact {
  // Name of the artifact
  name: String!
  // Name of the extension that produced this extension
  provider: String!
  // Human-friendly description of the artifact
  description: String
  // Available versions of this artifact
  versions: [...ArtifactVersion!]!
+ // Platform for this artifact
+ platform: [...ArtifactPlatform!]!
  // Additional metadata
  labels: [...ArtifactLabel!]
}

+ type ArtifactPlatform {
+   // Architecture of the artifact
+   architecture: String!
+   // OS of the artifact
+   os: String!
+   // variant of the architectur of the artifact
+   variant: String
+}
```

We could reuse the `Platform` from the current API, but I believe we should split the field into this `ArtifactPlatform` proposal so we can use them as we need.
With a cross-compilation toolchain, a pipeline could generate artifacts for multiple platform (I believe).

Would the artifacts only be OCI compliant artifacts?

**Comment by sipsma** [comment]
Agree this would be great to have. Some scattered thoughts:

In terms of the API, I think it would be good to find a way to support `Container` in addition to `Directory` (maybe `File` too).

re: backend API, this could just be a special, simplified case of the existing support for converting code->graphql schema. You'd just look for any function signature where no args are present and a `Directory` (or `Container`, `File`) is returned:
```go
func MyCoolArtifact() (*dagger.Directory, error) {
   ...
}
```

That's pretty trivial to do each our our SDK langs.

It does bring up another question, which is whether we'd want artifacts to support taking args. It slightly increases the scope of what they are, but if we restricted the args to be simple types like `string`, `int`, etc. plus lists of those types, then I can't think of any reason why it would be bad to support that.

**Comment by sipsma** [comment]
> What if the dagger engine could act as an "virtual registry", where all OCI artifacts could be docker-pulled directly?

Also, yeah that seems completely feasible and useful. I guess it would make the most sense for `Container` artifacts, but you could also just generate a default image config for `Directory`/`File` artifacts

**Comment by shykes** [comment]
> In terms of the API, I think it would be good to find a way to support `Container` in addition to `Directory` (maybe `File` too).

I was wondering about that. It seems to me we have two options:

1. We make container images a special case. After all, there is a special `Container` type (insert @vito terminology bikeshedding taunt here 😘 ), Dagger is incredibly good at producing that particular kind of artifact, building container images *is* a common use case, and there is the coolness of the "virtual registry" feature, which only works for container images. But, doesn't it result in every other type of artifacts being second-class citizens?

2. We make all artifacts equal: OCI image is just one type of artifact, it could be anything. But in that case: how do we fit the `Container` type and "virtual registry" idea? And how do we categorize all the different types of artifacts that could exist?3. 

> re: backend API, this could just be a special, simplified case of the existing support for converting code->graphql schema. You'd just look for any function signature where no args are present and a `Directory` (or `Container`, `File`) is returned:
> 
> ```go
> func MyCoolArtifact() (*dagger.Directory, error) {
>    ...
> }
> ```
> 
> That's pretty trivial to do each our our SDK langs.
> 
> It does bring up another question, which is whether we'd want artifacts to support taking args. It slightly increases the scope of what they are, but if we restricted the args to be simple types like `string`, `int`, etc. plus lists of those types, then I can't think of any reason why it would be bad to support that.

One benefit of artifacts as a first application of the extension system (which is what this is, as you know :), is that we can get away with a lot of restrictions on the schema. For example, I don't think we even need to expose a concept of "arguments", even with a restricted type system. I think we could restrict even further, and only expose a name, version, and a map of string labels.

In other words:


```go
func MyCoolArtifact(ctx context.Context, name, version string, labels map[string]string) (*dagger.Directory, error) {
   ...
}
```

**Comment by shykes** [comment]
> 2. We make all artifacts equal: OCI image is just one type of artifact, it could be anything. But in that case: how do we fit the `Container` type and "virtual registry" idea? And how do we categorize all the different types of artifacts that could exist?

Attempting to answer my own question here: what if we piggy-backed on OCI's attempt to generalize the spec to other types of artifacts, beyond container images? Presumably they have done work to categorize content types, and support their discovery and download from a registry. So perhaps we could piggy-back on that entire spec: OCI's nomenclature of artifact types, is our nomenclature of artifact types; etc.

**Comment by shykes** [comment]
> [...] what if we piggy-backed on OCI's attempt to generalize the spec to other types of artifacts, beyond container images? 

Some more research in that direction:

* [ORAS project](https://oras.land) ("OCI Registry As Storage")
* [OCI guidelines for artifact support](https://github.com/opencontainers/artifacts/blob/main/artifact-authors.md)
* [Me asking Twitter about this](https://twitter.com/solomonstre/status/1644122105322278912)

**Comment by shykes** [comment]
TIL [homebrew uses OCI for artifact storage!](https://twitter.com/zhangjintao9020/status/1644137888060145666)
